packages/flutter/test/foundation/assertions_test.dart

```diff
@@ -428,4 +428,367 @@ void main() {
     expect(trace, isNotNull);
     expect(trace.value, stack);
   });
+
+  test('defaultStackFilter compresses huge ugly stacks', () {
+    expect(
+      FlutterError.defaultStackFilter(hugeUglyStack.split('\n')),
+      const <String>[
+        '#2      new Text (package:flutter/src/widgets/text.dart:287:10)',
+        '#3      _MyHomePageState.build (package:assertions/main.dart:100:15)',
+        '#4      StatefulElement.build (package:flutter/src/widgets/framework.dart:4414:27)',
+        '#5      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4303:15)',
+        '#6      Element.rebuild (package:flutter/src/widgets/framework.dart:4027:5)',
+        '#7      ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4286:5)',
+        '#8      StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:4461:11)',
+        '#9      ComponentElement.mount (package:flutter/src/widgets/framework.dart:4281:5)',
+        '#348    _Timer._runTimers (dart:isolate-patch/timer_impl.dart:384:19)',
+        '#349    _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:418:5)',
+        '#350    _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:174:12)',
+        '(elided 330 frames from package flutter)'
```

Comment:
here in particular i think we are eliding way too much valuable information. It'd be better if between `#9` and `#348` we said something like "Inflated 28 elements". Even more than that though, because consider frames `#21` or `#60` for example, notice how those are different from the rest of the pattern.

I definitely wouldn't elide frames `#333` to `#339` inclusive. I think on the other hand we _should_ elide `#348` to `#350`, that we don't is just an oversight in the current code.

BTW we should make these tests verify the output on real stacks generated by the test, because we want to catch future regressions where Dart or the Flutter framework change structure and the logic here bitrots again.

Reply:
I don't think we can properly handle user code this way, but if it's a custom widget we wouldn't omit it here because that would live outside of package:flutter.

Part of the problem is a stack trace line just doesn't give us enough information (particularly on web, where we don't even get the class name). What might be possible is iterating from the bottom up until we hit a mount call from framework.dart, then group those together until we get to the one before we hit user code (and repeat if there's a custom element in user code).

Reply:
It's fine if we do a bad job on Web because Web doesn't give us enough yet. We should aim for doing the best we can on each platform we support, and not get tied down by the least common denominator.

Reply:
we could have pattern matchers for specific stack patterns we know about. if we do that, then there should be a way for these patterns to get registered by the bindings, so that each layer can register its own matchers.